src_base/xeve_args.c

```diff
@@ -0,0 +1,2313 @@
+
+        //     n += snprintf(allowed_values+n,  ARGS_OPT_ALLOWED_VALUES_MAXLEN, "\n          allowed values: [");
+        //     while(enums[i]!=ENUM_END) {
+        //         n += snprintf(allowed_values+n,  ARGS_OPT_ALLOWED_VALUES_MAXLEN, "%d, ", args_opt_table[idx].allowed_values.enums[i]);
+        //         i++;
+        //     }
+        //     n += snprintf(allowed_values+n-2,  ARGS_OPT_ALLOWED_VALUES_MAXLEN, "]");
+        //     break;
+        // case OPT_TYPE_CONST:
+        //     constants = args_opt_table[idx].allowed_values.constants;
+        //     while(constants[i].name) {
+        //         n += snprintf(allowed_values + n,  ARGS_OPT_ALLOWED_VALUES_MAXLEN, "\n          - %d: %s",args_opt_table[idx].allowed_values.constants[i].value, args_opt_table[idx].allowed_values.constants[i].name);
+        //         i++;
+        //     }
+        //     break;
+        case VAL_TYPE_STRING:
+            strings = args_opt_table[idx].allowed_values.list.strings;
+            while(strings[i].value) {
+                n += snprintf(allowed_values + n,  ARGS_OPT_ALLOWED_VALUES_MAXLEN, "\n          - %s",args_opt_table[idx].allowed_values.list.strings[i].value);
+                i++;
+            }
+            break;
+        default:
+            break;
+    }
+
+    if(opt->key != ARGS_NO_KEY)
+    {
+        snprintf(*help, ARGS_OPT_HELP_STRING_MAXLEN, "  -%c, --%s [%s]%s%s\n    : %s %s", opt->key, opt->key_long,
+                vtype, (optional) ? " (optional)" : "", (optional) ? default_value : "", opt->desc, allowed_values);
+    }
+    else
+    {
+        snprintf(*help, ARGS_OPT_HELP_STRING_MAXLEN, "  --%s [%s]%s%s\n    : %s %s", opt->key_long,
+                vtype, (optional) ? " (optional)" : "", (optional) ? default_value : "", opt->desc, allowed_values);
+    }
+
+
+
+    return XEVE_OK;
+}
+
+int xeve_args_num_options(ARGS_PARSER * args)
+{
+    if(!args) {
+        return 0;
+    }
+    return args->num_option;
+}
+
+/* Three-part iterator */
+char* xeve_args_first_key(ARGS_PARSER * args)             /* returns smallest element in n */
+{
+    ARGS_OPT *opt = NULL;
+
+    // do some checks
+    if( !args )
+        return NULL;
+
+    if(!args->num_option) return NULL;
+
+    opt = args->opts;
+    return opt->key_long;
+}
+
+// int nums_done(ARGS_PARSER * args, char* key);        /* returns 1 if key is past end */
+char* xeve_args_next_key(ARGS_PARSER * args, char* keyl)   /* returns next value after key */
+{
+    ARGS_OPT *opt = NULL;
+    int idx = 0;
+    int next_idx = 0;
+
+    // do some checks
+    if( !args)
+        return NULL;
+
+    if(keyl)
+    {
+        idx = args_search_long_key(args->opts, keyl);
+        if( idx  < 0 ) return NULL;
+
+        next_idx = idx + 1;
+
+        // if next idx is out of bound
+        if( next_idx > args->num_option-1) return NULL;
+
+
+    } else {
+        next_idx = 0;
+    }
+
+    opt = args->opts + next_idx;
+
+    return opt->key_long;
+}
```

Comment:
is this functions is still required?

Reply:
Yes, it is. xeve_args_get_option_description(ARGS_PARSER * args, char* keyl, char ** help) function needs the key value to provide a description for a given option related to that key. And to get the subsequent keys the
xeve_args_next_key() function is needed. It makes it able to get all subsequent keys for all options.